[file-format] add friend visibility

mengxu-fb opened the pull request:
<!--
Thank you for sending a PR. We appreciate you spending time to help improve the Diem project.

The project is undergoing daily changes. Pull Requests will be reviewed and responded to as time permits.
-->

## Motivation

This PR aims to add finer-grained visibility control to the Move language on top of the existing `public` and `private` visibility modifier. To be specific, given a function `f`, three access modifiers can be specified at function definition:
- **public**: functions defined in any module / script can call function `f`
- **private**: only functions defined in the same module can call function `f`
- **[IN THIS PR] public(friend)**: functions defined in the same module + *friend modules* (i.e., a controlled list of entities) can call function `f`

### Added language syntax and semantics:

There will be two additions added to the language in the first stage of implementation:
1. A friend list, specifiable at each module, that lists the modules that are considered as a *friend* of the host module.
    - Currently, only modules *within the same address as the host module* can be added to the friend list.
2. A protected keyword, specifiable at each function, to indicate the function can be callable from "friend" modules.

In summary, any function defined in the modules in the **friend list** can call any **protected** function defined in the host module.

In the following example:

```
0x2 {
  module A {
    friend 0x2::B;                      --> A friend list declared for each module
    friend 0x2::C;

    public(friend) fun foo() {}         --> A new visibility modifier: public(friend)
    public(friend) fun bar() {}

    fun i_am_private() {}
    public fun i_am_public() {}
  }
}
```

- Any function defined in host module (`0x2::A`) or friend modules (`0x2::B`, `0x2::C`) can call `foo` and `bar`, but no other functions can.
- Only functions defined in `0x2::A` can call `i_am_private`.
- Any function / script can call `i_am_public`.
- Only modules owned by `0x2` can be placed in the friend list.

### Progress tracking

This PR will be confined to the file format + basic checks (e.g., bounds check and duplication check) associated with the new file format. Other TODO items will come in follow-up PRs.

- Bytecode format
  - [DONE] versioned deserializer
  - [DONE] file format changes
  - [DONE] serialization of the new format with visibility info
  - [DONE] deserialization of both old and new versions of the bytecode
  - [DONE] valid friend list generation strategy (for proptest)
  - [TODO] more tests, especially on differences of the two versions of bytecode format
- Bytecode verifier
  - [DONE] bounds check of the new `CompiledModule` + tests + invalid mutations
  - [DONE] duplication check on the friend list + tests
  - [DONE] dependency check on whether a callsite can call into a `public(friend)` function
  - [TODO] existence of friend modules
  - [TODO] validity of friend declaration (e.g., creating cyclic dependencies is not allowed)
  - [TODO] same-address rule in friend declaration
  - [TODO] module upgrade compatibility checking
- IR compiler
  - [TODO] syntax and parsing for friend list and the protected modifier
  - [TODO] validation and dependency passing
  - [TODO] test cases on the new visibility feature
- Source compiler
  - [DONE] interface file generation with `public(friend)` functions
  - [TODO] syntax and parsing for friend list and the protected modifier
  - [TODO] validation and dependency passing
  - [TODO] test cases on the new visibility feature
- Move prover
  - [TODO] changes to `move-model`
- Diem framework
  - [TODO] re-write some of the modules with the new *protected* keyword
- Workflow
  - [WIP] finalize on the module publishing / re-publishing flow
  - [WIP] fix remaining issues related single-module republishing
  - [TODO] single-signer multi-module publishing in one transaction 
  - [TODO] E2E test cases

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

`cargo xtest`

New test cases will be added and existing tests will be updated per each commit. We will try to keep each commit small and sensible while maintaining a passing status of each commit.

## Related PRs

This PR depends on #7323 authored by @tnowacki
Another related development effort is VM republishing flow, tracked by #7143 and #7234.

mengxu-fb commented on language/testing-infra/e2e-tests/goldens/language_e2e_testsuite::tests::data_store::borrow_after_move.exp:
```diff
@@ -1,4 +1,4 @@
-Ok([TransactionOutput { write_set: WriteSet(WriteSetMut { write_set: [(AccessPath { address: 5455a8193f1dfc1b113ecc54d067afe1, path: 01000000000000000000000000000000010b4469656d4163636f756e740b4469656d4163636f756e7400 }, Value(20f5b9d6f01a99e74c790e2f330c092fa05455a8193f1dfc1b113ecc54d067afe1015455a8193f1dfc1b113ecc54d067afe1015455a8193f1dfc1b113ecc54d067afe100000000000000001800000000000000005455 ...
+Ok([TransactionOutput { write_set: WriteSet(WriteSetMut { write_set: [(AccessPath { address: 5455a8193f1dfc1b113ecc54d067afe1, path: 01000000000000000000000000000000010b4469656d4163636f756e740b4469656d4163636f756e7400 }, Value(20f5b9d6f01a99e74c790e2f330c092fa05455a8193f1dfc1b113ecc54d067afe1015455a8193f1dfc1b113ecc54d067afe1015455a8193f1dfc1b113ecc54d067afe100000000000000001800000000000000005455 ...
```
E2E tests golden files updated because the version number in the `CompiledModule` is bumped from 1 to 2.

mengxu-fb replied:
With 13661c2, not only the version number is bumped to 2, the `flags` byte is changed as well.

tnowacki commented on language/vm/src/file_format.rs:
```diff
@@ -338,8 +338,37 @@ pub struct FieldInstantiation {
+    }
+}
+
+/// A `FriendDeclaration` encapsulates information about a friend.
+#[derive(Clone, Debug, Hash, Eq, PartialEq)]
+#[cfg_attr(any(test, feature = "fuzzing"), derive(Arbitrary))]
+#[cfg_attr(any(test, feature = "fuzzing"), proptest(no_params))]
+pub struct FriendDeclaration {
+    /// Handle index to the friend module
+    pub module: ModuleHandleIndex,
+    // TODO: for future extensions, we may want to attach policies to friend declarations, e.g.,
+    // in the case of cross-address friendship, we may have a policy controlling whether the callee
+    // needs to sign if the caller wants to make an update.
+}
```
We can probably remove this struct?

tnowacki replied:
I'm guessing you added this because of the TODO of more policies, but there isn't really any point to having this struct now. And we will have to bump the fileformat version and do a refactor for any extension 

So best to not have the layer of indirection here if we don't need it

mengxu-fb replied:
Yes, the main reason for a struct here is for the ease of extensibility. But yes, the struct is not really needed now.

tnowacki replied:
Yeah I don't see it making it easier to making things extensible in the future. As in, the amount of work to actually extend anything isn't reduced by this struct

mengxu-fb replied:
After some refactoring trials I still want to keep the struct even if there is only one field here. It kinds of gives me a mentality that I am dealing with a "friend" instead of just a ModuleId, and we just happen to define that the unit of "friend" is a module.

But I also see the point that it may be good to ease the concept of "friend" out from the bytecode level, especially since we haven't fully decided on the name, so the fewer "friend" in var/struct names, the better.

Tough choice.

tnowacki commented on language/vm/src/file_format.rs:
```diff
@@ -1615,6 +1646,9 @@ pub struct CompiledModuleMut {
     pub struct_defs: Vec<StructDefinition>,
     /// Function defined in this module.
     pub function_defs: Vec<FunctionDefinition>,
+
+    /// Friend declarations of this module.
+    pub friend_decls: Vec<FriendDeclaration>,
```
I'm not super convinced that we should be using the term "friend" any more. I think at the bytecode level we should just be more explicit and descriptive. Pulling into a single term like "friend" seems useful only if we need to refer to this in a bunch of different places. Which I don't think is true?

So perhaps just 
```
pub protected_access: Vec<ModuleHandleIndex>
```

mengxu-fb replied:
I am fine with either naming, but I guess we may actually need refer to the "friend" concept across the stack? like file format, bytecode verifier, IR compiler, source compiler, prover, VM loader, etc?

sblackshear replied:
I strongly agree--I think `friend` is a bad term for us to use because it implies symmetry, but the language feature is not symmetric. Perhaps we should elevate this terminology question to a broader forum, but I think something like "client" or "patron" that connotes anti-symmetry would be good.

tnowacki replied:
Agree with Sam, and I would go one step further: I think we don't need to pin ourselves to a term like "friend" or "client" at all because the usecase is narrow. 

My vote for the syntax is 
```
protected  { 0x1::A, 0x2::B, 0x3::C }
```
No need for another keyword and it makes it hopefully more clear what is going on. If I saw "friend" here I would have no initial guess that it relates to `protected`

mengxu-fb replied:
After some internal discussion, we decided to go with the `friend` keyword and using `public(friend)` as the visibility modifier.

tnowacki replied:
Do we want `friend` for `friends`? :P 

Seriously, I don't know. 

I think `public(friend)` reads better. But `friends { 0x1::A, 0x2::B }` reads better

mengxu-fb replied:
I think `public(friend)` reads better and aligns with `public(script)`.

An alternative of friend declaration would be
```
friend 0x1::A;
friend 0x2::B;
```
similar to how we deal with `use`.

bob-wilson replied:
I'll vote for the plural. `public(friends)` reads better to me. If you compare it to Rust's `pub(crate)`, the code is only in one crate so it makes sense to use singular `crate` -- the visibility is public within the one crate. In our case, `public(friends)` means that the function is public to friend modules, and since there can be multiple friends, `public(friends)` is the right way to say that.

bob-wilson replied:
...but after reading Meng's response, I would also feel OK about the singular `friend`. The comparison to `public(script)` is better than my comparison to Rust's `pub(crate)`

tnowacki replied:
AH! very clever! I like it :) 

We should maybe add coalescing to this and to use. So you can do `friend 0x2::{A, B};`

We should do this for use too. But its okay if it doesn't make it in the first PR

tnowacki reviewed (commented):
In this "progress tracking" you have a big section for each major component. Do you plan on doing all of that on this PR? (please don't :))

sblackshear commented on language/vm/src/file_format.rs:
```diff
@@ -338,8 +338,37 @@ pub struct FieldInstantiation {
+#[derive(Clone, Debug, Eq, PartialEq)]
+#[cfg_attr(any(test, feature = "fuzzing"), derive(Arbitrary))]
+#[cfg_attr(any(test, feature = "fuzzing"), proptest(no_params))]
+pub enum Visibility {
```
I think it is redundant/error-prone to have both this enum and the `u8` constants. I would kill the `enum` and stick to the constants, since that's the style used elsewhere in the file format. The reason it is done this way: using raw bytecodes makes it easier to see what the binary format is, whereas using Rust-specific features like enums can obscure that.

mengxu-fb replied:
Not sure I entirely get this. When speaking about the file format, there may be two things we are talking about:
1. The `CompiledModule`, which is known in the Rust world. In `CompiledModule`, the `Visibility` is an enum,
2. The serialized bytes, which is what is stored on-chain. In the raw bytes, the `Visibility` is a bit in the u8 flag.

A similar situation is the handling of `Kind {All, Resource, Copyable}`. It is an enum in the Rust world, and a `u8` flag in the serialized bytes.

tnowacki replied:
I partially agree, and ran into a similar issue with constraints (working on an initial PR for that but was blocked by the versioning thing)

The benefits for the enum here are huge for writing code that checks visibility, granted that code is really rather small. 

I think you could maybe do 
```
#[repr(u8)]
pub enum Visibility {
  Private = 0x0,
  Public = 0x1,
  Protected = 0x2,
}
```
And then use these in the serializer and deserializer. 

So I guess I agree with sam that it is bad to have an enum and the u8 constants, but it is okay if the enum is the same as the u8 constants. It's when you have two things that things go wrong.

tnowacki replied:
The flags case is a bit different, but this is how we represent bytecodes (with an repr'd enum and then another bag of opcode constants)

tnowacki replied:
Yeah I think deserializing flags into enums is probably fine, especially since they are ensured to be exclusive in the deserializer, putting it not into an enum would feel misleading. 

But I think where `Kind` didn't do the best was having the u8 representation totally different from the enum

mengxu-fb replied:
Got it! Thanks for the description and the suggestion! I think we can go with this:

```
#[repr(u8)]
pub enum Visibility {
  Private = 0x0,
  Public = 0x1,
  Protected = 0x2,
}
```

tnowacki replied:
I wouldn't get too eager and switch to that just yet :) I'd like to see if that jives with @sblackshear's concerns

sblackshear replied:
I think the enum + `repr` is a great approach. My main concern was indeed about the redundant representation.

mengxu-fb replied:
Updated with 13661c2. Now the bytecode format is a bit more different:

Originally, the `flags` byte has two bits in use:
- 0x1: is_public
- 0x2: is_native

With the new format, the `flags` byte now essentially becomes two bytes:
- byte 1: visibility (with 0 - private, 1 - public, 2 - protected, and in the future, 3 - script, 4 - package, etc...).
- byte 2: now only bit 0x2 is used to indicate whether this is_native.

mengxu-fb commented:
> In this "progress tracking" you have a big section for each major component. Do you plan on doing all of that on this PR?

Actually, I am not sure what is the best way forward:
1) having a separate commit for each functionality but group all commits under this same PR
2) having smaller PRs (like this one, it is self-contained, and all tests are passing), and merge them separately.

Thoughts?

sblackshear commented:
I think (2) is much easier to review + is less likely to result in merge conflicts.

tnowacki commented on language/vm/src/deserializer.rs:
```diff
@@ -1107,13 +1205,41 @@ fn load_field_instantiations(
-    let is_public = (flags & FunctionDefinition::PUBLIC) != 0;
+
+    let visibility = match cursor.version() {
+        VERSION_1 => {
```
Probably fine to just use 1 and 2 here. I don't see the constant doing much :)

mengxu-fb replied:
To me having a constant imported from somewhere feels a bit safer... But yeah, purely personal preference.

tnowacki replied:
I'm not sure how much safer it is if VERSION_1 is always just 1?

Maybe it conveys a bit more intent? I could see that maybe, but I think `version() < 2` has been fairly clear in my usecase

mengxu-fb replied:
Yeah, I think VERSION_1 is probably going to be 1 all the time, and same for VERSION_2. Maybe the usecase I was thinking about is, say, I want to look at the places where there is a special handling of VERSION_n, I could do a `grep VERSION_n` to pinpoint the places.

tnowacki replied:
Yeah thats not a bad one. Let's do the constants then

tnowacki commented on language/vm/src/deserializer.rs:
```diff
@@ -1107,13 +1205,41 @@ fn load_field_instantiations(
+            }
+        }
+        VERSION_2 => {
+            // NOTE: changes compared with VERSION_1
+            // - in VERSION_1: flag bit 0x4 can be either 0 or 1, with no impact on deserialization,
+            // - in VERSION_2: flag bit 0x4 can only be 1 when bit 0x1 is 0, malformed otherwise.
+            if (flags & FunctionDefinition::PUBLIC) != 0 {
+                if (flags & FunctionDefinition::PROTECTED) != 0 {
+                    return Err(PartialVMError::new(StatusCode::MALFORMED).with_message(
+                        "A function cannot be both public and protected".to_string(),
+                    ));
+                }
+                Visibility::Public
+            } else if (flags & FunctionDefinition::PROTECTED) != 0 {
+                Visibility::Protected
+            } else {
+                Visibility::Private
+            }
```
I would rewrite this as 
```
let is_protected = (flags & FunctionDefinition::PROTECTED) != 0;
let is_public = (flags & FunctionDefinition::PUBLIC) != 0;
match (is_protected, is_public) {
  (false, false) => Private,
  (true, false) => Protected,
  (false, true) => Public,
  (true, true0 => malformed
}
```
I find this easier to case through than nested ifs

tnowacki commented:
Definitely (2). 

Luckily you can stub a lot of things out with protected as any other code can just fail when it encounters it for now

mengxu-fb commented:
Cool, option (2) -- small and incremental PRs it is then :)

mengxu-fb commented:
Rebased past #7534, merge conflicts with script visibility resolved.

mengxu-fb commented:
I addressed most comments in this PR except for whether we want to keep the single-field struct `FriendDeclaration`.

```rust
pub struct FriendDeclaration {
    /// Handle index to the friend module
    pub module: ModuleHandleIndex,
}
```

After some refactoring trials I still want to keep the struct even if there is only one field here. It kinds of gives me a mentality that I am dealing with a "friend" instead of just a `ModuleId`, and we just happen to define that the unit of "friend" is a module.

tnowacki commented on language/bytecode-verifier/src/dependencies.rs:
```diff
@@ -88,13 +91,21 @@ impl<'a, 'b> Context<'a, 'b> {
+                let may_be_called = match func_def.visibility {
+                    Visibility::Public | Visibility::Script => true,
+                    Visibility::Friend => {
+                        if let Some(self_module_id) = &self_module {
+                            friend_module_ids.contains(self_module_id)
+                        } else {
+                            false
+                        }
```
if-let + else is not my friend :P 

Either switch to a match, or `self_module.map(|self_id| friend_module_ids.contains(self_id)).unwrap_or(false)`

tnowacki commented on language/compiler/ir-to-bytecode/src/compiler.rs:
```diff
@@ -565,6 +565,8 @@ pub fn compile_module<'a>(
         constant_pool,
         struct_defs,
         function_defs,
+        // TODO
```
Maybe tag yourself in it? Not needed, but I find that helpful sometimes. I do `// TODO(tmn)`

Regardless, a short description is nice incase this is missed and someone comes and reads the code later

mengxu-fb replied:
It's actually already done, but yes, I can elaborate the TODO and tag myself.

tnowacki commented on language/move-model/src/model.rs:
```diff
@@ -2398,6 +2398,11 @@ impl<'env> FunctionEnv<'env> {
             || match self.definition_view().visibility() {
                 Visibility::Public => true,
                 Visibility::Script => panic!("Script visibility not yet supported"),
+                // TODO: for simplicity, we may treat `public(friend)` the same way as `public`
+                // based on the fact that they both expose the function to external modules. We may
+                // want to change the function name from `is_public` to `is_exposed` or have
+                // iner-grained return values (instead of just a bool) to indicate visibility.
+                Visibility::Friend => panic!("Friend visibility not yet supported"),
```
nit, I might have done this for the script case... but you can do `unimplemented!(..)` which gives a *slightly* better message when it is hit

tnowacki commented on language/vm/src/file_format.rs:
```diff
@@ -338,8 +338,17 @@ pub struct FieldInstantiation {
 
-/// A `StructDefinition` is a type definition. It either indicates it is native or
-// defines all the user-specified fields declared on the type.
+/// A `FriendDeclaration` encapsulates information about a friend.
+#[derive(Clone, Debug, Hash, Eq, PartialEq)]
+#[cfg_attr(any(test, feature = "fuzzing"), derive(Arbitrary))]
+#[cfg_attr(any(test, feature = "fuzzing"), proptest(no_params))]
+pub struct FriendDeclaration {
+    /// Handle index to the friend module
+    pub module: ModuleHandleIndex,
+}
+
```
We should remove this struct as the encapsulation is not needed

mengxu-fb replied:
`FriendDeclaration` dropped.

tnowacki commented on language/vm/src/file_format.rs:
```diff
@@ -1642,6 +1657,9 @@ pub struct CompiledModuleMut {
     pub struct_defs: Vec<StructDefinition>,
     /// Function defined in this module.
     pub function_defs: Vec<FunctionDefinition>,
+
+    /// Friend declarations of this module.
+    pub friend_decls: Vec<FriendDeclaration>,
```
Nit, I'd move this up with the other tables. It is weird for it to come after them (probably put it right after field handles)

mengxu-fb replied:
Since we are moving this, I think we can put it right after `module_handles` and before  `struct_handles`? The rationale is that the friend list is expected to be declared before any functions and structs are defined.

mengxu-fb replied:
Ignore the comment above. Yes, this list should be after the handles. After `field_handles` is the right place.

tnowacki commented on language/move-vm/runtime/src/unit_tests/vm_arguments_tests.rs:
```diff
@@ -204,6 +204,8 @@ fn make_module_with_function(
                 code: vec![Bytecode::LdU64(0), Bytecode::Abort],
             }),
         }],
+
+        friend_decls: vec![],
```
be sure to move this when you move the field in the declaration order

tnowacki commented on language/vm/src/compatibility.rs:
```diff
@@ -87,8 +87,12 @@ impl Compatibility {
         // old module's public functions are a subset of the new module's public functions
-        for function in &old_module.externally_visible_functions {
-            if !new_module.externally_visible_functions.contains(&function) {
+        //
+        // TODO: the current implementation treats `public(friend)` function the same way as`public`
+        // for compatibility checking. We might need to change this in the future to allow
+        // `public(friend)` functions to update if none of its friends actually calls it.
+        for item in &old_module.externally_visible_functions {
+            if !new_module.externally_visible_functions.contains(&item) {
```
We should totally rewrite this. I can put up another PR after this lands as you work on other things

mengxu-fb replied:
I have re-written this in https://github.com/diem/diem/pull/7531/commits/d518d19e907ea64e5ca36ae615824f1c03e84186.

tnowacki commented on language/vm/src/deserializer.rs:
```diff
@@ -508,6 +508,15 @@ fn build_common_tables(
             | TableType::FIELD_INST => continue,
+            TableType::FRIEND_DECLS => {
+                // friend declarations do not exist in VERSION_1
+                if binary.version() == VERSION_1 {
```
if binary.version() < Version_2 ?

tnowacki commented on language/vm/src/deserializer.rs:
```diff
@@ -1162,6 +1175,24 @@ fn load_struct_definition_indices(
+        .checked_add(table.count as usize)
+        .expect("Unexpected overflow as the error should be detected early by `check_tables`");
+    let mut cursor = binary.new_cursor(start, end);
+    while cursor.position() < u64::from(table.count) {
```
I imagine Rust will CSE this, but maybe  move `u64::from(table.count)` to a local binding? Ignore this if none of the other `load_<table>` code does this

mengxu-fb replied:
Other `load_*` seem to be doing this as well, like `load_identifiers` and `load_constant_pool`. So I guess this should be OK.

tnowacki replied:
I imagine Rust is smart enough :) Just something to consider

tnowacki commented on language/vm/src/file_format.rs:
```diff
@@ -1590,6 +1603,8 @@ impl CompiledScriptMut {
 
             struct_defs: vec![],
             function_defs: vec![main_def],
+
+            friend_decls: vec![],
```
be sure to move this when you move the field in the declaration order

tnowacki commented on language/vm/src/file_format.rs:
```diff
@@ -1720,6 +1738,7 @@ impl Arbitrary for CompiledModuleMut {
                 vec(any::<StructDefinition>(), 0..=size),
                 vec(any_with::<FunctionDefinition>(size), 0..=size),
             ),
+            vec(any::<FriendDeclaration>(), 0..=size),
```
be sure to move this when you move the field in the declaration order

tnowacki reviewed (approved):
Thanks for all the iterations! Looks good!

mengxu-fb commented:
/land
